### DISCUSSION TITLE
How to extend model & override findFirst()

### DISCUSSION THREAD

**Discussion by quasipickle** [phalcon/cphalcon#15514] [Q&A] [answered]
I'm migrating from 3.4 to 4.  I have a model that is extending \Phalcon\Mvc\Model and overriding `findFirst()` - simply to put a caching layer in.

v4 introduced the fact that `findFirst()` _must_ return null or an object of type \Phalcon\Mvc\ModelInterface.  But how does that work if I've set `columns` or `hydration` parameters?  If I use the built-in method it works fine, but if I use my extended method I get an exception saying "**Return value of Model\ModelBase::findFirst() must implement interface Phalcon\Mvc\ModelInterface or be null, instance of Phalcon\Mvc\Model\Row returned**".  

All I'm doing is forwarding along the result of `parent::findFirst()` - I don't know what else to do to allow my overriding function to properly return arrays or Model\Row objects.

Thanks.

**Comment by rudiservo** [comment, marked as answer]
I would use a query builder honestly.
We can either consider this a bug or we accept that since it is in v4 this a breaking change so v4 is actually ok with the versioning.
The manual however is not ok since it says it can be done.

IMO hydration into array should not be used directly the model::find, it promotes inconsistency and it might break a lot of things, also adds a lot more checks just to handle those cases, not only consuming more cycles but also giving it more room to generate problems.
Things should be has statically typed has possible, it's easier on codding, IDE, internal checks, etc.

**Reply by quasipickle** [reply]
I guess you're right - breaking changes are allowed between major versions.  I think the easiest solution is just to never specify columns or hydration.  Using the QueryBuilder isn't really an option here, as this functionality lives in my base model, not an individual model.  To use QueryBuilder I'd have to have knowledge of what model I'm actually acting on, as well as pull out all the parameters.  Thanks for the suggestion though.

I'll just retrieve full objects from now on.

**Reply by rudiservo** [reply]
you can always make your own function and just put in whatever you need in it.
This is the current findFirst in Zephir, you can easily make it in php and create you own ::hydrateFindFirst($parameters)
```
public static function findFirst(var parameters = null) -> <ModelInterface> | null
    {
        var params, query;

        if null === parameters {
            let params = [];
        } elseif typeof parameters === "array" {
            let params = parameters;
        } elseif typeof parameters === "string" || is_numeric(parameters) {
            let params = [parameters];
        } else {
            throw new Exception(
                "Parameters passed must be of type array, string, numeric or null"
            );
        }

        let query = static::getPreparedQuery(params, 1);

        /**
         * Return only the first row
         */
        query->setUniqueRow(true);

        /**
         * Execute the query passing the bind-params and casting-types
         */
        return query->execute();
    }
```

**Reply by quasipickle** [reply]
Thanks for the advice.  I'd rather not go digging through the codebase everywhere though.  Honestly I don't think the uses of hydration and columns is all that widespread, and I think the caching functionality gives more of a performance boost than hydrating as arrays does.
